pkg/cluster/clusterserviceprincipal.go

```diff
@@ -173,50 +177,50 @@ func (m *manager) updateAROSecret(ctx context.Context) error {
 
 func (m *manager) updateOpenShiftSecret(ctx context.Context) error {
-	var changed bool
-	spp := m.doc.OpenShiftCluster.Properties.ServicePrincipalProfile
-	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
-		//data:
-		// azure_client_id: secret_id
-		// azure_client_secret: secret_value
-		// azure_tenant_id: tenant_id
-		secret, err := m.kubernetescli.CoreV1().Secrets("kube-system").Get(ctx, "azure-credentials", metav1.GetOptions{})
-		if err != nil {
-			return err
-		}
-
-		if string(secret.Data["azure_client_id"]) != spp.ClientID {
-			secret.Data["azure_client_id"] = []byte(spp.ClientID)
-			changed = true
-		}
-
-		if string(secret.Data["azure_client_secret"]) != string(spp.ClientSecret) {
-			secret.Data["azure_client_secret"] = []byte(spp.ClientSecret)
-			changed = true
-		}
-
-		if string(secret.Data["azure_tenant_id"]) != m.subscriptionDoc.Subscription.Properties.TenantID {
-			secret.Data["azure_tenant_id"] = []byte(m.subscriptionDoc.Subscription.Properties.TenantID)
-			changed = true
-		}
-
-		if changed {
-			_, err = m.kubernetescli.CoreV1().Secrets("kube-system").Update(ctx, secret, metav1.UpdateOptions{})
-			if err != nil {
-				return err
-			}
-		}
-		return nil
-	})
-	if err != nil {
+	//data:
+	// azure_client_id: secret_id
+	// azure_client_secret: secret_value
+	// azure_tenant_id: tenant_id
+	secret, err := m.kubernetescli.CoreV1().Secrets(clusterauthorizer.AzureCredentialSecretNameSpace).Get(ctx, clusterauthorizer.AzureCredentialSecretName, metav1.GetOptions{})
+	if kerrors.IsNotFound(err) {
+		secret = &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      clusterauthorizer.AzureCredentialSecretName,
+				Namespace: clusterauthorizer.AzureCredentialSecretNameSpace,
+			},
+			Data: map[string][]byte{},
+		}
+	} else if err != nil {
 		return err
 	}
 
+	resourceGroupID := m.doc.OpenShiftCluster.Properties.ClusterProfile.ResourceGroupID
+	spp := m.doc.OpenShiftCluster.Properties.ServicePrincipalProfile
+	desiredData := map[string][]byte{
+		"azure_subscription_id": []byte(m.subscriptionDoc.ID),
+		"azure_resource_prefix": []byte(m.doc.OpenShiftCluster.Properties.InfraID),
+		"azure_resourcegroup":   []byte(resourceGroupID[strings.LastIndex(resourceGroupID, "/")+1:]),
+		"azure_region":          []byte(m.doc.OpenShiftCluster.Location),
+		"azure_client_id":       []byte(spp.ClientID),
+		"azure_client_secret":   []byte(spp.ClientSecret),
+		"azure_tenant_id":       []byte(m.subscriptionDoc.Subscription.Properties.TenantID),
+	}
+
 	// return early if not changed
+	var changed bool
+	for key, val := range desiredData {
+		changed = bytes.Compare(val, secret.Data[key]) != 0 || changed
+	}
 	if !changed {
 		return nil
 	}
 
+	secretApplyConfig := applyv1.Secret(secret.Name, secret.Namespace).WithData(desiredData)
+	_, err = m.kubernetescli.CoreV1().Secrets(clusterauthorizer.AzureCredentialSecretNameSpace).Apply(ctx, secretApplyConfig, metav1.ApplyOptions{FieldManager: "aro-rp", Force: true})
```

Comment:
There are two new fields in `metav1.ApplyOptions{}`. `FieldManager` is metadata about the entity to last change the affected fields, in this case the RP. The `Force` parameter is required for use in controllers ([ref]()https://kubernetes.io/blog/2022/10/20/advanced-server-side-apply/#controller-authors).

Reply:
If we plan to re-use server side apply with the force flag elsewhere in the codebase, we may want to declare `FieldManager` somewhere else rather than hardcoding it so that we can re-use this value.

I am imagining a scenario where we use server side apply with this resource somewhere else in the codebase, but `FieldManager: "aro-rp"` is set differently (maybe `FieldManager: "rp"`) and we get a conflict.

It doesn't seem to matter as long as we use force: true, but maybe we could use some global solution.

Reply:
> I am imagining a scenario where we use server side apply with this resource somewhere else in the codebase, but FieldManager: "aro-rp" is set differently (maybe FieldManager: "rp") and we get a conflict.

That's exactly the reason you want to use the force flag in a controller. From what I understood from the documents, the FieldManager exists to tell user-run apply operations that the field was set by a different entity and the user could then decide how to handle that information, but when using it in some sort of controller (which this essentially is, just one that's triggered manually) you want to just apply your information. We could even set different FieldManagers depending on the RP operation that's setting the field, which may be more useful for traceability. This one could be something like `"az-aro-update"`.